activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/helper/ErrorPropagation.java

```diff
@@ -276,65 +276,98 @@ protected static Map<String, List<Event>> findCatchingEventsForProcess(
-                String eventErrorCode = retrieveErrorCode(bpmnModel, errorEventDef.getErrorRef());
-
-                if (eventErrorCode == null || compareErrorCode == null || eventErrorCode.equals(compareErrorCode)) {
-                    List<Event> elementBoundaryEvents = null;
-                    if (!eventMap.containsKey(boundaryEvent.getAttachedToRefId())) {
-                        elementBoundaryEvents = new ArrayList<Event>();
-                        eventMap.put(boundaryEvent.getAttachedToRefId(), elementBoundaryEvents);
-                    } else {
-                        elementBoundaryEvents = eventMap.get(boundaryEvent.getAttachedToRefId());
+            if (boundaryEvent.getAttachedToRefId() != null) {
+                ErrorEventDefinition errorEventDef = getErrorEventDefinition(boundaryEvent);
+                if (errorEventDef != null) {
+                    String eventErrorCode = retrieveErrorCode(bpmnModel, errorEventDef.getErrorRef());
+
+                    if (isErrorCodeMatching(eventErrorCode, compareErrorCode)) {
+                        // Separate boundary events with no error code (catch-all) to be processed last
+                        if (eventErrorCode == null) {
+                            boundaryEventsWithoutErrorCode.add(boundaryEvent);
+                        } else {
+                            addBoundaryEventToMap(eventMap, boundaryEvent);
+                        }
                     }
-                    elementBoundaryEvents.add(boundaryEvent);
                 }
             }
         }
+
+        // Second pass: Add boundary events WITHOUT error codes (catch-all) at the end
+        for (BoundaryEvent boundaryEvent : boundaryEventsWithoutErrorCode) {
+            addBoundaryEventToMap(eventMap, boundaryEvent);
+        }
+
         return eventMap;
     }
 
+    /**
+     * Helper method to add a boundary event to the event map.
+     * Uses computeIfAbsent for cleaner code (Java 8+).
+     *
+     * @param eventMap the map to add the event to
+     * @param boundaryEvent the boundary event to add
+     */
+    private static void addBoundaryEventToMap(Map<String, List<Event>> eventMap, BoundaryEvent boundaryEvent) {
+        eventMap.computeIfAbsent(boundaryEvent.getAttachedToRefId(), k -> new ArrayList<>()).add(boundaryEvent);
+    }
+
+    /**
+     * Extracts the ErrorEventDefinition from an event's event definitions list.
+     *
+     * @param event the event to extract from
+     * @return the ErrorEventDefinition if found, null otherwise
+     */
+    private static ErrorEventDefinition getErrorEventDefinition(Event event) {
+        if (CollectionUtil.isNotEmpty(event.getEventDefinitions()) &&
+            event.getEventDefinitions().get(0) instanceof ErrorEventDefinition) {
+            return (ErrorEventDefinition) event.getEventDefinitions().get(0);
+        }
+        return null;
+    }
```

Comment:
I think this method can be added directly to Event class (maybe returning an `Optional<ErrorEventDefinition>`), so it can also be reused in other places where we need to retrieve the error event definition from the event. Then we can simply call `event.getErrrorEventDefinition()`. WDTY?